Towards #56 - Update end-to-end-gentle.md

juanpablovs opened the pull request:
Towards #56 .

Changes proposed in this PR:

- Gentle introduction example is homogenized with main2
- Added the file that will be submitted to the judges
- Added "setup" and "what the judges will do"

trentmc commented:
> Fixes #56 .

Please change PR description and title to have "Towards #56"

More comments to follow...

trentmc commented:
My [feedback](https://github.com/oceanprotocol/predict-eth/pull/59#issuecomment-1331959588) for [PR#59](  https://github.com/oceanprotocol/predict-eth/pull/59) also applies to this PR.

I'll summarize:
- **Key thing: test data needs to be from 24h, 23h, .., 12h in the past. And your model can only learn from data older than that. The README needs to get fixed accordingly.**
- it should be obvious to readers to see how this README expands upon the steps in main2.md. Therefore, this README's section numbers should correspond to section numbers in main2.md.

trentmc reviewed (changes requested):
Changes requested: see [previous comment](https://github.com/oceanprotocol/predict-eth/pull/58#issuecomment-1331965193)

juanpablovs commented:
I took care of your comments, in particular:

1. test data needs to be from 24h -- done
2. readme sections correspond with main2 -- done
3. rewrote section 3.2 and now computes nmse with the 12 next subsequent values as per main2 -> 3.3
4. removed ancillary information from the example, such as select data per specific hour
5. rewrote text significantly (mostly at the end) to accomodate these changes

thanks,

trentmc commented on examples/end-to-end-gentle.md:
```diff
@@ -304,48 +288,89 @@ X = base[['AAPL_Open', 'AAPL_High', 'AAPL_Low', 'AAPL_Close', 'AAPL_Vol', 'GOOG_
+# We use the last 12 hours for testing, all the rest of data is for training
+
+# Target y_test and y_train
+y_test = y.iloc[0:12]
```
Slices for train and test data need to be flipped.

Data is ordered from oldest to newest, and right now it's selecting the oldest data as the data to predict.

And, be sure that this section reconciles well with my main comment below: need to properly predict 12 hours into the future.

juanpablovs replied:
not sure I understand -- we sorted the dataframe above, in the line (# Sort the values by date and reset the indexes of the DataFrame:)
Now row 0 of the data frame is the most recent date (2022-11-04 20:00:00) and testing takes the 12 first values. I just re-checked and it does take the 12 most recent ones. That is the appropriate data for testing, right? (for the record; oldest data is 2022-08-09 05:00:00)

May be you prefer not to sort the dataframe and reverse the testing/training variables? if that's the case let me know and I can do it in no time

trentmc replied:
> testing takes the 12 first values

This is a time series. Data is ordered by time. If testing takes the first 12 values, then it's training on future data to predict past values.

juanpablovs replied:
for you "first" is row 0 or the last row?

trentmc replied:
>> Assuming it's sorted as "oldest first"
> for you "first" is row 0 or the last row?
First row

However you do it: make sure that the test data is the newest (most recent) values.

juanpablovs replied:
Please note that this entire comment is based on you assuming that data is sorted as "oldest first".
Important to remember that we make assumptions when something is unverifiable or too expensive/time consuming etc to verify
This is not the case.
You have the code and the dataset. No need to make the assumption. It really takes less than 1 min to verify that data is not sorted as "oldest first" -- which is what ignited this discussion
However -- as instructed -- I will verify again that the data is properly sorted.

juanpablovs replied:
Please note that this entire comment is based on you assuming that data is sorted as "oldest first".
Important to remember that we make assumptions when something is unverifiable or too expensive/time consuming etc to verify
This is not the case.
You have the code and the dataset. No need to make that assumption. It really takes less than 1 min to verify that data is not sorted as "oldest first" -- which is what ignited this discussion
However -- as instructed -- I will verify again that the data is properly sorted.

trentmc replied:
> May be you prefer not to sort the dataframe and reverse the testing/training variables? if that's the case let me know and I can do it in no time -- this is probably my fault as I usually place the most recent value on top

Yes please make sure that row 0 is oldest, row 1 is next-oldest, ..., row n-2 is second-newest, row n-1 is newest. Thank you.

juanpablovs commented on examples/end-to-end-gentle.md:
```diff
@@ -304,48 +288,89 @@ X = base[['AAPL_Open', 'AAPL_High', 'AAPL_Low', 'AAPL_Close', 'AAPL_Vol', 'GOOG_
-However, wait a second, wait a second, what about if instead of using all the data we deploy or model for only one hour of the day, say 10am?
+Same as we did before, with ccxt we can get the next 12 real prices of ETH
+
+Don't forget to import the nmse function. [It's in the Setup in the 'helpers' section](https://github.com/oceanprotocol/predict-eth/blob/main/support/helpers.md)
```
done

juanpablovs commented on examples/end-to-end-gentle.md:
```diff
@@ -304,48 +288,89 @@ X = base[['AAPL_Open', 'AAPL_High', 'AAPL_Low', 'AAPL_Close', 'AAPL_Vol', 'GOOG_
-
-r2 = r2_score(y_test, predictions) # result is 0.68
-nmse = calc_nmse(y_test, predictions) # result is 0.0048
+# These are the next 12 predictions to submit to the judges
```
"Shouldn't have to slice predictions into pred_vals, since there should be just 12 predicted values"
The previous model shuffled data for fitting with a sample train size of 30%. Since now the predictions train size is fixed what you see here is from the past model. I have removed the line accordingly because now the array size will always be 12

"But, then, macro: is this truly getting you the next 12 hours?"
Yes, more information here: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression.predict
in a nutshell: by passing to predict the 12 X_test values we get their corresponding 12 ETH predictions.

About your suggestions - This example is for teaching purposes, we try to lure people in, nobody is asking for sophisticated math

trentmc replied:
Every readme needs approach the problem in a conceptually correct fashion. Otherwise something incorrect is being taught. I am concerned that this readme is not quite there yet. Maybe it is, maybe not. That's why I am asking questions.

A related thing that I just noticed (sorry for not noticing earlier) - regarding this code
```python
# Add predictor variables
X = base[['AAPL_Open', 'AAPL_High', 'AAPL_Low', 'AAPL_Close', 'AAPL_Vol', 'GOOG_Open', 'GOOG_High', 'GOOG_Low', 'GOOG_Close', 'GOOG_Vol', 'TSLA_Open', 'TSLA_High', 'TSLA_Low', 'TSLA_Close', 'TSLA_Vol', 'AMZN_Open', 'AMZN_High', 'AMZN_Low', 'AMZN_Close', 'AMZN_Vol']]

# Add the target
y = base['ETH_Price']
```

For this code, does this mean that the model does not take into account past values of ETH when making its prediction? But rather just values of ETH, GOOG, etc. To be precise: if it's Monday at 7.00am, and I want to predict the price Monday at 8.00am, will the model be using the values of ETH for 7.00am, 6.00am, 5.00am etc as input? 

To extend the example: it's still Monday at 7.00am. I want to predict the price Monday at 8.00am, _and_ at 9.00am. How does it do that right now?

trentmc replied:
> For this code, does this mean that the model does not take into account past values of ETH when making its prediction? But rather just values of ETH, GOOG, etc. To be precise: if it's Monday at 7.00am, and I want to predict the price Monday at 8.00am, will the model be using the values of ETH for 7.00am, 6.00am, 5.00am etc as input?
> To extend the example: it's still Monday at 7.00am. I want to predict the price Monday at 8.00am, and at 9.00am. How does it do that right now?

This has not been resolved at all.

juanpablovs commented on examples/end-to-end-gentle.md:
````diff
@@ -304,48 +288,89 @@ X = base[['AAPL_Open', 'AAPL_High', 'AAPL_Low', 'AAPL_Close', 'AAPL_Vol', 'GOOG_
+Don't forget to import the nmse function. [It's in the Setup in the 'helpers' section](https://github.com/oceanprotocol/predict-eth/blob/main/support/helpers.md)
 
 ```python
+# To start we get the most recent date from our base dataset
````
I think I misunderstood what had to be done here and I produced the next 12 subsequent values.
I rewrote 3.2 to only compute the nmse with the real and expected values

juanpablovs commented on examples/end-to-end-gentle.md:
```diff
@@ -304,48 +288,89 @@ X = base[['AAPL_Open', 'AAPL_High', 'AAPL_Low', 'AAPL_Close', 'AAPL_Vol', 'GOOG_
 
-# repeat the previous model here
+# With ccxt we get the next 12 real values of the target ETH price
+cex_vals_array = ccxt.binance().fetch_ohlcv('ETH/USDT', '1h', since=initial_time_timestamp, limit=12)
```
see previous comment

trentmc commented on examples/end-to-end-gentle.md:
```diff
@@ -4,7 +4,7 @@ This is an example full submission to the challenge. You can use it as starting
 
 Remember:
```
In the comments below, you'd requested that I go through the README myself. I agree with this as a goal. In practice my time is stretched extremely thin, so most of the time I follow an 80/20 approach: review by inspecting the code and giving feedback that way.

Given the request, I took the time to go walk through the README. The user flow is disjointed in sections 1 and 2, with gaps in what needs to be done. We need to fix that. Here's specific feedback:

### Section 1 Feedback

This section currently has the user install from command line -> import within python -> install more in command line -> create a test py file -> do more python things. This won't work because the "import within python" will be lost. Second, given the goal to homogenize with main2.md, using a test py file departs too much. Third, small thing, this end-to-end readme doesn't need to install ccxt since main2.md already does. Therefore, section 1 flow needs to become:

1. From challenge 2 setup, do section 1.1, 1.2, 1.3 (stuff prior to python)
2. In this readme, pip install pandas, ccxt, scikit-learn
3. From challenge 2 setup, do section 1.4, 1.5
4. Then: "In the Python console, do: <#import libraries; import pandas as pd; etc>

### Section 2 Feedback

The section currently has 
> The data is available on Ocean market for free: https://market.oceanprotocol.com/asset/did:op:59e8ce4a599e7d3efc3f23ce68c6ae92b08e5764c0a8f801c47ae00c47db1522

Then it proceeds to:
```python
df_tech_apple = pd.read_csv('AAPL_1hr.csv', usecols=['time','open', 'high', 'low', 'close', 'volume']) #open csv
 ```

If the user tries that, they will get: 
```python
FileNotFoundError: [Errno 2] No such file or directory: 'AAPL_1hr.csv'
```

Why? Because there weren't any instructions to actually download the file 'AAPL_1hr.csv'.

What needs to be done here: have the Python code using Ocean library to download 'AAPL_1hr.csv'. [This README](https://github.com/oceanprotocol/predict-eth/blob/main/examples/get-ethdata-ocean-binance.md) shows how.

And the same needs to be done for Google, Tesla, and Amazon stock.

trentmc replied:
Section 1 of my feedback above is now resolved.

Section 2 is not:
- `FileNotFoundError: [Errno 2] No such file or directory: 'AAPL_1hr.csv'`

This is still not resolved.

trentmc reviewed (changes requested):
Several changes requested, and Q's asked. There will likely be more once this is done.

juanpablovs commented:
Hi Trent, I rewrote the example as instructed. Let me know .Thanks

trentmc commented on examples/end-to-end-gentle.md:
```diff
@@ -23,7 +23,7 @@ Ready?, let's start! :rocket: :rocket: :rocket:
 ---
 
-## 1. Understanding the Challenge
+## 0. Understanding the Challenge
```
Good to see. Also, there's a section before this that's un-numbered. Perhaps make both part of section 0. E.g. with subsection 0.1 and 0.2

trentmc commented on examples/end-to-end-gentle.md:
```diff
@@ -45,46 +45,37 @@ Can we do it? Let's give it a try by first gathering the data for our model.
+- [x] Setup 1.4 and 1.5
 
-Open the file. Add the following content to the file:
+Copy and paste everything below:
```
Make this more standard. Change to: "In the Python console:"
